File: src/detector/helperProtocol.ts
import { DetectorLogEntry, ImeSnapshot } from "../model/types";

interface JsonRecord {
  [key: string]: unknown;
}

export const PROTOCOL_VERSION = 1;

export interface HelloMessage {
  type: "hello";
  version: number;
  capabilities: string[];
}

export function parseHelloLine(line: string): HelloMessage | undefined {
  const parsed = parseJsonRecord(line);
  if (!parsed || parsed.type !== "hello") {
    return undefined;
  }

  if (typeof parsed.version !== "number") {
    return undefined;
  }

  const capabilities = Array.isArray(parsed.capabilities)
    ? parsed.capabilities.filter((value): value is string => typeof value === "string")
    : [];

  return {
    type: "hello",
    version: parsed.version,
    capabilities
  };
}

export function parseSnapshotLine(line: string): ImeSnapshot | undefined {
  const parsed = parseJsonRecord(line);
  if (!parsed || parsed.type !== "state") {
    return undefined;
  }

  const state =
    parsed.state === "cn" || parsed.state === "en" || parsed.state === "unknown"
      ? parsed.state
      : "unknown";

  return {
    type: "state",
    state,
    timestamp: typeof parsed.timestamp === "string" ? parsed.timestamp : new Date().toISOString(),
    source: "native-helper",
    imeName: typeof parsed.imeName === "string" ? parsed.imeName : undefined,
    isOpen: typeof parsed.isOpen === "boolean" ? parsed.isOpen : undefined,
    layoutHex: typeof parsed.layoutHex === "string" ? parsed.layoutHex : undefined,
    threadId: typeof parsed.threadId === "number" ? parsed.threadId : undefined,
    hwnd: typeof parsed.hwnd === "string" ? parsed.hwnd : undefined,
    reason: typeof parsed.reason === "string" ? parsed.reason : undefined,
    confidence: typeof parsed.confidence === "number" ? parsed.confidence : undefined,
    rawStateAvailable:
      typeof parsed.rawStateAvailable === "boolean" ? parsed.rawStateAvailable : undefined
  };
}

export function parseLogLine(line: string): DetectorLogEntry | undefined {
  const parsed = parseJsonRecord(line);
  if (!parsed || parsed.type !== "log" || typeof parsed.message !== "string") {
    return undefined;
  }

  return {
    type: "log",
    level: parsed.level === "error" || parsed.level === "warn" ? parsed.level : "info",
    message: parsed.message,
    timestamp: typeof parsed.timestamp === "string" ? parsed.timestamp : new Date().toISOString(),
    details: parsed.details,
    source: typeof parsed.source === "string" ? parsed.source : "native-helper"
  };
}

function parseJsonRecord(line: string): JsonRecord | undefined {
  try {
    const parsed = JSON.parse(line) as unknown;
    if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
      return undefined;
    }

    return parsed as JsonRecord;
  } catch {
    return undefined;
  }
}

File: jetbrains/src/main/kotlin/com/chestnutch/cursorimehud/protocol/HelperProtocol.kt
package com.chestnutch.cursorimehud.protocol

import com.chestnutch.cursorimehud.model.DetectorLogEntry
import com.chestnutch.cursorimehud.model.HelloMessage
import com.chestnutch.cursorimehud.model.ImeSnapshot
import com.chestnutch.cursorimehud.model.ImeState
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import java.time.Instant

const val PROTOCOL_VERSION: Int = 1
const val MAX_LINE_BYTES: Int = 64 * 1024
const val MAX_BUFFER_BYTES: Int = 1024 * 1024

object HelperProtocol {
  fun parseHelloLine(line: String): HelloMessage? {
    val record = parseRecord(line) ?: return null
    if (record.stringOrNull("type") != "hello") return null
    val version = record.intOrNull("version") ?: return null
    val capabilities = record.get("capabilities")
      ?.takeIf { it.isJsonArray }
      ?.asJsonArray
      ?.mapNotNull { if (it.isJsonPrimitive && it.asJsonPrimitive.isString) it.asString else null }
      ?: emptyList()

    return HelloMessage(version, capabilities)
  }

  fun parseSnapshotLine(line: String): ImeSnapshot? {
    val record = parseRecord(line) ?: return null
    if (record.stringOrNull("type") != "state") return null
    val state = ImeState.fromWire(record.stringOrNull("state")) ?: return null

    return ImeSnapshot(
      state = state,
      timestamp = record.stringOrNull("timestamp") ?: Instant.now().toString(),
      imeName = record.stringOrNull("imeName"),
      isOpen = record.booleanOrNull("isOpen"),
      layoutHex = record.stringOrNull("layoutHex"),
      threadId = record.numberOrNull("threadId")?.toLong(),
      hwnd = record.stringOrNull("hwnd"),
      reason = record.stringOrNull("reason"),
      confidence = record.numberOrNull("confidence")?.toDouble(),
      rawStateAvailable = record.booleanOrNull("rawStateAvailable")
    )
  }

  fun parseLogLine(line: String): DetectorLogEntry? {
    val record = parseRecord(line) ?: return null
    if (record.stringOrNull("type") != "log") return null
    val message = record.stringOrNull("message") ?: return null
    val level = when (record.stringOrNull("level")) {
      "error" -> "error"
      "warn" -> "warn"
      else -> "info"
    }

    return DetectorLogEntry(
      level = level,
      message = message,
      timestamp = record.stringOrNull("timestamp") ?: Instant.now().toString(),
      source = record.stringOrNull("source") ?: "native-helper",
      details = record.get("details")?.toString()
    )
  }

  fun refreshCommand(): String = "{\"command\":\"refresh\"}\n"

  private fun parseRecord(line: String): JsonObject? = try {
    val parsed = JsonParser.parseString(line.trim())
    if (parsed.isJsonObject) parsed.asJsonObject else null
  } catch (_: Exception) {
    null
  }
}

private fun JsonObject.stringOrNull(name: String): String? {
  val value = get(name) ?: return null
  return if (value.isJsonPrimitive && value.asJsonPrimitive.isString) value.asString else null
}

private fun JsonObject.booleanOrNull(name: String): Boolean? {
  val value = get(name) ?: return null
  return if (value.isJsonPrimitive && value.asJsonPrimitive.isBoolean) value.asBoolean else null
}

private fun JsonObject.numberOrNull(name: String): Number? {
  val value = get(name) ?: return null
  return if (value.isJsonPrimitive && value.asJsonPrimitive.isNumber) value.asNumber else null
}

private fun JsonObject.intOrNull(name: String): Int? {
  val value = get(name) ?: return null
  if (!value.isJsonPrimitive || !value.asJsonPrimitive.isNumber) return null
  return value.asJsonPrimitive.asString.toIntOrNull()
}
